examples/time_series_smoother.py

```diff
@@ -0,0 +1,380 @@
+    )
+    test_dtl = TimeSeriesDataloader(
+        x_file="data/toy_time_series_smoother/x_test_sin_smoother.csv",
+        date_time_file="data/toy_time_series_smoother/x_test_sin_smoother_datetime.csv",
+        output_col=output_col,
+        input_seq_len=input_seq_len,
+        output_seq_len=output_seq_len,
+        num_features=num_features,
+        stride=seq_stride,
+        x_mean=train_dtl.x_mean,
+        x_std=train_dtl.x_std,
+        time_covariates=["hour_of_day", "day_of_week"],
+        keep_last_time_cov = True
+    )
+
+    # Viz
+    viz = PredictionViz(task_name="forecasting", data_name="sin_signal")
+
+    # Network
+    net = Sequential(
+        SLSTM(num_features + input_seq_len - 1, 40, 1),
+        SLSTM(40, 40, 1),
+        SLinear(40, 1),
+    )
+
+    # net.to_device("cuda")
+    net.set_threads(1)  # multi-processing is slow on a small net
+    net.input_state_update = True
+    out_updater = OutputUpdater(net.device)
+
+    # -------------------------------------------------------------------------#
+    # Training
+    mses = []
+    # Initialize the sequence length
+    mu_sq = np.ones(input_seq_len,dtype=np.float32)
+    pbar = tqdm(range(num_epochs), desc="Training Progress")
+
+    for epoch in pbar:
+        batch_iter = train_dtl.create_data_loader(batch_size, shuffle=False)
+
+        # Decaying observation's variance
+        sigma_v = exponential_scheduler(
+            curr_v=sigma_v, min_v=0.3, decaying_factor=0.99, curr_iter=epoch
+        )
+        var_y = np.full((batch_size * len(output_col),), sigma_v**2, dtype=np.float32)
+        y_train = []
+
+        # for x, y in batch_iter:
+        for idx_sample, (x, y) in enumerate(batch_iter):
+
+            # replace nan in input x by the lstm_prediction:
+            if idx_sample < input_seq_len + infer_seq_len_smoother:
+                nan_indices = np.where(np.isnan(x))[0]
+                x[nan_indices] = mu_sq[nan_indices]
+
+            # Feed forward
+            m_pred, _ = net(x)
+
+            # Update output layer
+            out_updater.update(
+                output_states=net.output_z_buffer,
+                mu_obs=y,
+                var_obs=var_y,
+                delta_states=net.input_delta_z_buffer,
+            )
+            # Feed backward
+            net.backward()
+            net.step()
+
+            # Training metric
+            pred = normalizer.unstandardize(
+                m_pred, train_dtl.x_mean[output_col], train_dtl.x_std[output_col]
+            )
+            y_train.append(y)
+            obs = normalizer.unstandardize(
+                y, train_dtl.x_mean[output_col], train_dtl.x_std[output_col]
+            )
+            mse = metric.mse(pred, obs)
+            mses.append(mse)
+
+
+            # Update mu_sq
+            mu_sq = np.append(mu_sq,m_pred)
+            mu_sq = mu_sq[-input_seq_len:]
+
+
+        # Smoother
+        mu_zo_smooth, var_zo_smooth = net.smoother()
```

Comment:
This is not gonna scale because if you time series length > 1e6, it is a game over!

Reply:
So far, I will keep is it as it is. possible that smoothing is not applicable for cases.